Step1 - 엔티티 매핑

lewisseo91 opened the pull request:
안녕하세요!
이번 step1을 진행하며 기존에는 jpa를 정말 툴로만 생각하고 겉 핥기만 했다는 생각이 많이 들었습니다.
현재는 학습 가이드라인에 있는 영한님 유튜브를 보면서 공부중인데 정말 즐겁네요!
리뷰 해 주셔서 감사합니다~!
진행하면서 궁금한 점이 있어 PR 올리면서 같이 올리게 되었습니다~

- Repository 어노테이션 없이 autowired를 쓰면 빈 주입이 되고 interface에 있는 repository 빈 등록이 자동으로 되는 것이 흥미로워서 조사하게 되었는데요. 이 부분은 Spring boot에서 제공하는 기능으로 이해를 했습니다만, Auditing을 적용하는 도중에 EnableJpaAuditing에 대해 궁금한 점이 생겼습니다. 
springbootApplication위에 EnableJpaAuditing 어노테이션 없이도 실행 시 ddl에 제가 원하는 BaseEntity에 있는 createdAt이나 modifiedAt이 잘 적용되어 있는 것 같더라구요~
테스트를 할 때의 DataJpaTest 어노테이션 속에도 EnableJpaAuditing 과 같은 것들이 없어서 의아함을 느끼게 되었습니다.
'EnableJpaAuditing 없이도 작동을 원하는 부분은 작동하는 것 같은데 어떤 이유로 필요할까'라는 생각을 하게 되었습니다.
혹시 EnableJpaAuditing 어노테이션 없이 어플리케이션을 구동하면 어떤 일이 일어나는 지 팁을 얻을 수 있을까요? 
repository는 의견이 많아서 찾았는데 enableJpaAuditing은 정보가 너무 없어서 도움 요청드립니다 ㅠ 
제가 참고해 본 링크입니다. 
[repository annotation](https://stackoverflow.com/questions/45730111/spring-data-repository-works-without-annotations)
[repository annotation2](https://stackoverflow.com/questions/42691697/using-repository-annotation-when-implementing-jparepostiory-in-spring)
[spring document](https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.sql)

> The Spring Framework provides extensive support for working with SQL databases, from direct JDBC access using JdbcTemplate to complete “object relational mapping” technologies such as Hibernate. Spring Data provides an additional level of functionality: creating Repository implementations directly from interfaces and using conventions to generate queries from your method names.

- 그리고 이전 미션 TDD를 진행하면서 테스트 메소드의 네이밍을 camelCase로 진행하라는 advice를 받은 적이 있는데요. 이번에 service안에 만들어져 있는 메소드를 보니 _ 이후에 한글이 적혀있는 경우가 많더라구요~ 
혹시 뒤에 한글이 오면 _를 써도 괜찮나요?
        
리뷰 해 주셔서 정말 감사드립니다. 좋은 하루 되시길 바라요~!

lewisseo91 committed:
feat(env): show-sql 설정

- 실행시 sql 보이는 설정 추가

lewisseo91 committed:
feat(docs): 관계 예상 그림 1차

- 관계 예상 그림 1차

lewisseo91 committed:
feat(User, Answer, Question, DeleteHistory): Entity 추가

Global
- 쓰지않는 setter 추가
- no arg constructor 필요한 곳에 추가(protected)
Base Entity 추가
- 생성시간, 수정시간 추가
User Entity 추가
- 컬럼 nullable, unique, length 설정
Answer Entity 추가
- 컬럼 lob, nullable 설정
Question Entity 추가
- 컬럼 lob, nullable, length 설정
DeleteHistory Entity 추가
- Enumerate enum type 추가

lewisseo91 committed:
feat(UserTest, AnswerTest, QuestionTest, DeleteHistoryTest): 기본 테스트 추가

Global
- 기본 CRUD 테스트 추가
UserTest 추가
- user userId로 찾기 테스트 추가
- user 중에 lewisseo로 아이디 하나 추가
AnswerTest 추가
- question id 기준으로 삭제되지않은 answer 목록 찾기 테스트 추가
- question id 기준으로 삭제된 answer 목록 확인 테스트 추가
- 삭제되지않은 answer 찾기 테스트 추가
- 삭제한 answer 확인 테스트 추가
QuestionTest 추가
- 삭제되지 않은 question 목록 찾기 테스트 추가
- 삭제되지 않은 question 하나 찾기 테스트 추가
- 삭제한 question 찾기 테스트 추가
DeleteHistoryTest 추가
- CRUD중 CD 만 테스트 추가

lewisseo91 committed:
feat(docs): 2차 관계 예상 그림 추가

테스트 기반으로 예상 그림 변경 후 추가 (언제든지 변경 가능)

lewisseo91 committed:
style(test): 변수명 actual -> 적절해 보이는 변수명으로 변경

actual -> user, answer, deleteHistory, question 으로 변경

lewisseo91 committed:
chore(docs): 줄 바꿈

그림이 좀 이상하게 올라가서 변경

lewisseo91 committed:
feat(BaseEntity): baseEntity abstract 으로 변경

baseEntity abstract 으로 변경

lewisseo91 commented:
안녕하세요~!
오늘 강의 듣던 중에 제가 질문 했던 부분의 많은 부분들이 강의를 듣고 나서 이해할 수 있는 부분임을 깨달았습니다. 
학습 테스트는 강의 끝나고 스스로 공부를 조금 더 한 뒤에 다시 재요청 드리겠습니다!
혹시나 리뷰 보실까봐 강의 쉬는 시간에 노트 남겨 놓습니다!
좋은 하루 되세요~

lewisseo91 committed:
feat(env): 변수 log 값 확인하기 위해 추가

변수 log 값 확인하기 위해 추가

lewisseo91 committed:
feat(data): 예제 데이터 삽입

예제 데이터 삽입

lewisseo91 committed:
feat(docs): 강의 정리 내용 및 팁 추가

강의 정리 내용 및 팁 추가

lewisseo91 committed:
feat(subway): domain, test 학습 테스트 추가

학습 테스트 추가
- Station <-> Line 추가
- Line <-> LineStation <-> Station 추가 (LineStation을 가시화 함으로써 Many To Many 를 푼 것)
- Member -> Favorite 추가
- Repo 추가
- StationTest, MemberTest, LineTest 테스트 추가

lewisseo91 committed:
feat(docs): 수업 내용 추가

- 연관관계 설정 부분 설명 추가

lewisseo91 committed:
feat(docs): test에 displayname 추가

test에 displayname 추가

slamdunk7575 commented on README.md:
```diff
@@ -0,0 +1,71 @@
+# qna 서비스 jpa
+
+## 관계 예상 그림
+
+- 2차 예상 (테스트 케이스 만든 후)
+  
+  ![jwp-qna drawio-v2](https://user-images.githubusercontent.com/17772475/141173769-0c66d797-aba8-45ae-9339-045e410c234a.png)
```
관계도까지 그려보시고 정리를 열심히 해주셨네요 💯

slamdunk7575 commented on src/main/java/qna/domain/BaseEntity.java:
```diff
@@ -0,0 +1,29 @@
+import javax.persistence.MappedSuperclass;
+import java.time.LocalDateTime;
+
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
+public abstract class BaseEntity {
+
+    @Column(nullable = false)
+    @CreatedDate
+    private LocalDateTime createdAt = LocalDateTime.now();
+    @LastModifiedDate
+    private LocalDateTime updatedAt;
```
`BaseEntity` 분리 잘 해주셨네요 👍

줄바꿈을 통해 가독성을 높여보면 좋겠어요 :)

lewisseo91 replied:
앗 이 부분은 제가 관리를 못했네요~! 리뷰 감사합니다!!

slamdunk7575 commented on src/main/java/qna/domain/BaseEntity.java:
```diff
@@ -0,0 +1,29 @@
+@EntityListeners(AuditingEntityListener.class)
+public abstract class BaseEntity {
+
+    @Column(nullable = false)
+    @CreatedDate
+    private LocalDateTime createdAt = LocalDateTime.now();
```
> Q. Repository 어노테이션 없이 autowired를 쓰면 빈 주입이 되고 interface에 있는 repository 빈 등록이 자동으로 되는 것이 흥미로워서 조사하게 되었는데요. 이 부분은 Spring boot에서 제공하는 기능으로 이해를 했습니다만, Auditing을 적용하는 도중에 EnableJpaAuditing에 대해 궁금한 점이 생겼습니다.
springbootApplication위에 EnableJpaAuditing 어노테이션 없이도 실행 시 ddl에 제가 원하는 BaseEntity에 있는 createdAt이나 modifiedAt이 잘 적용되어 있는 것 같더라구요~
테스트를 할 때의 DataJpaTest 어노테이션 속에도 EnableJpaAuditing 과 같은 것들이 없어서 의아함을 느끼게 되었습니다.
'EnableJpaAuditing 없이도 작동을 원하는 부분은 작동하는 것 같은데 어떤 이유로 필요할까'라는 생각을 하게 되었습니다.
혹시 EnableJpaAuditing 어노테이션 없이 어플리케이션을 구동하면 어떤 일이 일어나는 지 팁을 얻을 수 있을까요?
repository는 의견이 많아서 찾았는데 enableJpaAuditing은 정보가 너무 없어서 도움 요청드립니다 ㅠ


원인은 현재 `private LocalDateTime createdAt = LocalDateTime.now();` 으로
생성시간을 셋팅하고 있어서 에러가 나지 않은 것입니다. 
`@EnableJpaAuditing` 설정을 사용하지 않으면 `Auditing` 기능이 제대로 동작하지 않습니다.
한번 확인해보시면 좋을 것 같아요 :)

[Spring Data JPA에서 Auditing 사용하는 법](https://devlog-wjdrbs96.tistory.com/415)

(참고)
- `@DataJpaTest`로 테스트 하면 내장된 In-Memory DB를 사용하는데요. 
아래 설정을 주면 MySQL 문법 쿼리와 실제 DB로도 테스트를 할 수 있습니다.
```
spring.jpa.properties.hibernate.dialect=org.hibernate.dialect.MySQL57Dialect
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
```

- `application.properties` 관련해서 미션 진행하시면서 참고해보시면 도움이 되실것 같아요.

```
# SQL 확인
spring.jpa.properties.hibernate.show_sql=true
spring.jpa.properties.hibernate.format_sql=true
logging.level.org.hibernate.type.descriptor.sql=trace

# 외부 h2 DB 연동
spring.datasource.url=jdbc:h2:~/test;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE;MODE=MySQL
spring.datasource.username=sa
spring.h2.console.enabled=true
spring.jpa.properties.hibernate.dialect=org.hibernate.dialect.MySQL57Dialect
```

lewisseo91 replied:
감사합니다!! 바로 테스트 해 보겠습니다~

slamdunk7575 replied:
저는 질문과 답변을 같이 드리기 위해 글을 붙여넣는 것이라 다른 이유가 없으시다면 동현님은 코멘트만 달으셔도 됩니다.
오히려 동현님 코멘트가 잘 안보이게 되어서요 😄

slamdunk7575 replied:
저는 질문과 답변을 같이 드리기 위해 글을 붙여넣는 것이라 다른 이유가 없으시다면 동현님은 코멘트만 달으셔도 됩니다.
오히려 동현님 코멘트가 잘 안보이게 되어서요 😄

slamdunk7575 replied:
저는 질문과 답변을 같이 드리기 위해 글을 붙여넣는 것이라 다른 이유가 없으시다면 동현님은 코멘트만 달으셔도 됩니다.
오히려 동현님 코멘트가 잘 안보이게 되어서요 😄

slamdunk7575 commented on src/main/java/qna/domain/BaseEntity.java:
```diff
@@ -0,0 +1,29 @@
+    @LastModifiedDate
+    private LocalDateTime updatedAt;
+
+    public LocalDateTime getCreatedAt() {
+        return createdAt;
+    }
+
+    public LocalDateTime getUpdatedAt() {
+        return updatedAt;
+    }
```
사용하지 않는 코드는 제거하는게 좋겠네요.

lewisseo91 replied:
> 사용하지 않는 코드는 제거하는게 좋겠네요.

리뷰 감사합니다~!

slamdunk7575 commented on src/main/java/qna/domain/User.java:
```diff
@@ -2,15 +2,29 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
+    @Column(nullable = false, length = 20, unique = true)
     private String userId;
```
👍

slamdunk7575 commented on src/main/java/qna/domain/User.java:
```diff
@@ -2,15 +2,29 @@
+    @Column(length = 50)
     private String email;
 
     private User() {
```
엔티티를 Repository에서 가져와 생성할때 public, protected 기본 생성자가 필요합니다.
접근 제한자를 `protected` 변경하면 좋겠네요.

[JPA Entity의 기본 생성자 관련 예외 정리하기](https://wbluke.tistory.com/6)

lewisseo91 replied:
아 이 부분도 제가 놓쳐버렸네요 ㅠ 감사합니다!

slamdunk7575 commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -1,6 +1,89 @@
     public static final Question Q2 = new Question("title2", "contents2").writeBy(UserTest.SANJIGI);
+
+    @Autowired
+    private QuestionRepository questions;
```
크게 중요한 부분은 아니지만 `repository` 라는 `네이밍`을 주면 더 좋겠네요. 
`questions` 는 주로 List 형태를 떠올리게 되어서요 😄

lewisseo91 replied:
아 학습 테스트 시에 s는 교육적 허용이라고 했는데 제가 이 부분 수정하는 걸 또 놓쳤네요 감사합니다~!

slamdunk7575 commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,100 @@
+        );
+    }
+
+    @DisplayName("question id 기준으로 삭제되지않은 answer 목록 찾기")
+    @Test
+    void findByQuestionIdAndDeletedFalseTest() {
```
> Q. 그리고 이전 미션 TDD를 진행하면서 테스트 메소드의 네이밍을 camelCase로 진행하라는 advice를 받은 적이 있는데요. 이번에 service안에 만들어져 있는 메소드를 보니 _ 이후에 한글이 적혀있는 경우가 많더라구요~
혹시 뒤에 한글이 오면 _를 써도 괜찮나요?

테스트를 작성할때 테스트에 대한 설명을 작성하면 나중에 동현님이나 동료 개발자가 유지보수할 때 더욱 수월하겠죠.
이렇게 테스트에 설명을 작성하기 위해 주로 `@DisplayName` 을 사용하거나 `메소드명을 한글`로 작성합니다.
꼭 정답이 있는 것은 아니지만 한글을 사용하면 아무래도 IDE 호환도 좋지 않고 하여 저는 주로 `@DisplayName`을 선호하고 있습니다. 연습하는 과정이니 JUnit5에 추가된 `@DisplayName`을 사용하는 것도 좋을 것 같네요 🙂

lewisseo91 replied:
와우 제가 JUnit5 부터 했다보니 5에서 추가된 어노테이션이라는 것은 상상도 못했네요 리뷰 감사합니다!! 
displayName 잘 활용하겠습니다~

slamdunk7575 reviewed (changes requested):
안녕하세요~ 동현님. JPA 미션 함께하게 된 양경준 입니다.
Entity 맵핑 잘 해주셨네요. 질문 주신 부분에 대한 코멘트 남겨드렸는데 한번 확인해보시고 
재요청 주시면 좋을것 같아요~ 앞으로 미션도 잘부탁드립니다 💪

lewisseo91 committed:
feat(env): h2 외부 데이터 베이스 이용

h2 데이터 베이스 설치 후 테스트 진행
data.sql 삭제

lewisseo91 committed:
style(domain): Answer, BaseEntity, DeleteHistory.. 도메인 객체 줄 바꿈 조정

Answer, BaseEntity, DeleteHistory.. 도메인 객체 줄 바꿈 조정
필요없는 getter 삭제

lewisseo91 committed:
feat(test): 테스트 네이밍 -s -> -Repository 로 변경, createdAt, updatedAt 테스트 추가

테스트 네이밍 -s -> -Repository 로 변경
createdAt, updatedAt 테스트 추가
테스트에 EnableJpaAuditing, AutoConfigureTestDatabase 추가

lewisseo91 committed:
fix(User): private 접근자 -> protected 로 변경

private 접근자 -> protected 로 변경

lewisseo91 committed:
feat(DeleteHistory): DeleteHistory save 테스트를 위한 getter 추가

DeleteHistory save 테스트를 위한 getter 추가

lewisseo91 commented:
안녕하세요~! 
[Spring Data JPA에서 Auditing 사용하는 법](https://wbluke.tistory.com/6) 글이 정말 좋았습니다!
좋은 링크 감사합니다~! ListenerClass를 좀 더 자세히 이해했었어야 했구나 하고 깨달았습니다.
나머지 피드백 부분도 반영하고 외부 h2 테스트도 해 보았습니다~!
![image](https://user-images.githubusercontent.com/17772475/141458823-3e28565f-3715-4c01-9e60-017b080ea2b7.png)
이렇게 결과도 확인 해 보면서 진행하니 신기했습니다. 
좋은 피드백 정말 정말 감사합니다!

줄 바꿈 쪽에 하나 궁금한 것이 있어 질문 남깁니다~!
제가 어노테이션이 붙어있지 않은 멤버들도 한 줄 띄웠는데 괜찮을까요?
다른 코드들 중에 보니 어노테이션이 붙지 않으면 붙여놓는 것 같은데 
붙여놓았더니 저는 오히려 읽기가 어려워져서 한번 떨어뜨려 놓아 보았습니다.
![image](https://user-images.githubusercontent.com/17772475/141459745-f45e68a5-aeb3-42a5-a660-f74f2b1f2a32.png)

리뷰 해 주셔서 감사합니다~!
잘 부탁드립니다~

아 그리고 제가 답변이 너무 좋아서 블로그에 기록하려고 하는데 괜찮을까요?

slamdunk7575 commented on src/main/java/subway/Application.java:
```diff
@@ -0,0 +1,11 @@
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+public class Application {
```
이곳에 `@EnableJpaAuditing` 을 추가하여 `Auditing`을 활성화하면 테스트에는 추가하지 않아도 될것 같네요.

lewisseo91 replied:
> 이곳에 `@EnableJpaAuditing` 을 추가하여 `Auditing`을 활성화하면 테스트에는 추가하지 않아도 될것 같네요.

좋은 피드백 감사합니다~! 다음 step 초기 단계에 해당 부분은 수정하겠습니다~!

slamdunk7575 commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -1,22 +1,61 @@
+@Entity
+@Table(name = "delete_history")
 public class DeleteHistory {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
+
+    @Enumerated(EnumType.STRING)
     private ContentType contentType;
+
     private Long contentId;
+
     private Long deletedById;
+
     private LocalDateTime createDate = LocalDateTime.now();
```
> 줄 바꿈 쪽에 하나 궁금한 것이 있어 질문 남깁니다~!
제가 어노테이션이 붙어있지 않은 멤버들도 한 줄 띄웠는데 괜찮을까요?
다른 코드들 중에 보니 어노테이션이 붙지 않으면 붙여놓는 것 같은데
붙여놓았더니 저는 오히려 읽기가 어려워져서 한번 떨어뜨려 놓아 보았습니다.

저도 통일성과 가독성을 위해 띄어놓는 것을 선호합니다 :)

lewisseo91 replied:
오 자신이 없었다가 자신을 얻었습니다!
감사합니다 ^^!!

slamdunk7575 commented:
> 아 그리고 제가 답변이 너무 좋아서 블로그에 기록하려고 하는데 괜찮을까요?

`강의 자료나 내용을 그대로 올리는 경우엔 저작권 문제가 있을 수 있으니 참고해주세요 🙏`

동현님이 공부하신 것을 정리하여 올리는 것은 괜찮습니다 :)

slamdunk7575 reviewed (approved):
피드백 반영 잘해주셨네요 👍 
질문에 대한 답변과 한가지 코멘트 남겨드렸는데 확인해보시면 좋을것 같아요.
앞으로 미션도 화이팅 입니다~

lewisseo91 commented:
> > 아 그리고 제가 답변이 너무 좋아서 블로그에 기록하려고 하는데 괜찮을까요?
> 
> `강의 자료나 내용을 그대로 올리는 경우엔 저작권 문제가 있을 수 있으니 참고해주세요 🙏`
> 
> 동현님이 공부하신 것을 정리하여 올리는 것은 괜찮습니다 :)

감사합니다~!!
